ic-agent/src/agent/mod.rs

```diff
@@ -782,6 +863,31 @@ impl Agent {
             signed_request_status,
         })
     }
+
+    async fn get_subnet_by_canister(
+        &self,
+        canister: &Principal,
+    ) -> Result<Arc<Subnet>, AgentError> {
+        let subnet = self
+            .subnet_key_cache
+            .lock()
+            .unwrap()
+            .get_subnet_by_canister(canister);
+        if let Some(subnet) = subnet {
+            Ok(subnet)
+        } else {
+            let cert = self
+                .read_state_raw(vec![vec!["subnet".into()]], *canister)
+                .await?;
```

Comment:
Usually, the canister ranges check is performed within check_delegation.
However, if the certificate is from the root subnet, the delegation is absent. In the context of replica-signed queries, we need to perform extra check of the canister ranges for the root subnet, .
```
if cert.delegation.is_none(){
    let canister_range = lookup_value(&cert.tree, canister_range_lookup)?;
    let ranges: Vec<(Principal, Principal)> =
        serde_cbor::from_slice(canister_range).map_err(AgentError::InvalidCborData)?;
    if !principal_is_within_ranges(&effective_canister_id, &ranges[..]) {
        // the certificate is not authorized to answer calls for this canister
        return Err(AgentError::CertificateNotAuthorized());
    }
}
                
```

Reply:
The design slides and @roelstorms 's [comments](https://docs.google.com/presentation/d/1xbobmOhplo-OwjP41JY5MhpI7c-peyJPtNE-v_-DJKo/edit?disco=AAAA-gP82Ik) may be referenced for more context.

Reply:
I'm wondering if we only need to do this in case of RSQ:

> In the context of replica-signed queries, we need to perform extra check of the canister ranges for the root subnet,

I'd argue that this might be a general issue that a node on the root subnet could respond with the root cert (and thus root state tree) while requesting the state for a canister on another subnet.

In our case the issue is that we have the list of authorative nodes in that state and the root subnet nodes shouldn't be authoritive for RSQ on other subnets. But maybe there are other scenario's where this could be an issue. 

Is it worth mitigating this in verify(). In case there is no delegation, the effective canister id needs to be within the range of root subnet. I'm not sure if the root subnet also publishes it's own canister ranges in it's own certificate?

Reply:
> I'm not sure if the root subnet also publishes it's own canister ranges in it's own certificate

It is a special case for replica-signed queries where agents request the `/subnet` path and thus it contains `/subnet/<subnet-id>/canister_ranges` in the cert directly. 

In other cases of read_state call for other paths, `/subnet/<subnet-id>/canister_ranges` may not be present in the cert and thus moving the logic to `verify()` may not work.